certora/specs/DoubleEndedQueue.spec

```
@@ -0,0 +1,361 @@
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule clear {
+    clear@withrevert();
+    
+    // liveness
+    assert !lastReverted, "never reverts";
+    
+    // effect
+    assert length() == 0, "sets length to 0";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: front/back access reverts only if the queue is empty or querying out of bounds                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule onlyEmptyRevert(env e) {
+    require nonpayable(e);
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    method f;
+    calldataarg args;
+
+    bool emptyBefore = empty();
+
+    f@withrevert(e, args);
+
+    assert lastReverted => (
+        (f.selector == front().selector && emptyBefore) ||
+        (f.selector == back().selector && emptyBefore) ||
+        (f.selector == popFront().selector  && emptyBefore) ||
+        (f.selector == popBack().selector  && emptyBefore) ||
+        f.selector == at_(uint256).selector
+    ), "only revert if empty or out of bounds";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: at(key) only reverts if key is out of bounds                                                                  |
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule onlyOutOfBoundsRevert(uint256 key) {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    at_@withrevert(key);
+
+    assert lastReverted <=> key >= length(), "only reverts if key is out of bounds";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: only clear/push/pop operations can change the length of the queue                                             │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule noLengthChange(env e) {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    method f;
+    calldataarg args;
+
+    uint256 lengthBefore = length();
+    f(e, args);
+    uint256 lengthAfter = length();
+
+    assert lengthAfter != lengthBefore => (
+        (f.selector == pushFront(bytes32).selector && lengthAfter == lengthBefore + 1) ||
+        (f.selector == pushBack(bytes32).selector && lengthAfter == lengthBefore + 1) ||
+        (f.selector == popBack().selector && lengthAfter == lengthBefore - 1) ||
+        (f.selector == popFront().selector && lengthAfter == lengthBefore - 1) ||
+        (f.selector == clear().selector && lengthAfter == 0)
```

The length postconditions are already checked in each correctness rule.

```suggestion
        f.selector == pushFront(bytes32).selector ||
        f.selector == pushBack(bytes32).selector ||
        f.selector == popBack().selector ||
        f.selector == popFront().selector ||
        f.selector == clear().selector
```

---

That was sort of the previous layout but Hadrien suggested the current pattern, so that the length is associated with the function that causes it. I think it makes sense to have both although it's implicit in each function correctness check. What do you think?

---

I don't know, I would avoid the redundancies, given that verification is slow and expensive.

---

> I don't know, I would avoid the redundancies, given that verification is slow and expensive.

In this case, the check is done on values that are already computed/explored. So its doesn't really increase the complexity.

Equivalent in solidity would be doing a require on things that are already in the stack (cheap) vs doing a require on data that we need to read from storage and process.

Also, this entire spec file takes ~30sec. So its not really a concern here.

---

Yes, there's no perceivable difference when testing one vs the other, but I can run a benchmark if you like.

Also, I think it's valuable as a programmer to look at the verification rules and get insights from them. Although these rules are redundant, it's easier to see the length changes instead of looking for answers on other rules.